sklearn/preprocessing/target.py

```diff
@@ -0,0 +1,232 @@
+        a 2-dimensional array. If ``inverse_func is ``None``, the function used
+        will be the identity function. The inverse function is used to return
+        to the same space of the original training labels during prediction.
+
+    check_inverse : bool, (default=True)
+        Whether to check that ``transform`` followed by ``inverse_transform``
+        or ``func`` followed by ``inverse_func`` leads to the original targets.
+
+    Attributes
+    ----------
+    regressor_ : object
+        Fitted regressor.
+
+    transformer_ : object
+        Transformer used in ``fit`` and ``predict``.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from sklearn.linear_model import LinearRegression
+    >>> from sklearn.preprocessing import TransformTargetRegressor
+    >>> tt = TransformTargetRegressor(regressor=LinearRegression(),
+    ...                               func=np.log, inverse_func=np.exp)
+    >>> X = np.arange(4).reshape(-1, 1)
+    >>> y = np.exp(2 * X).ravel()
+    >>> tt.fit(X, y)
+    ... # doctest: +NORMALIZE_WHITESPACE
+    TransformTargetRegressor(check_inverse=True,
+                             func=<ufunc 'log'>,
+                             inverse_func=<ufunc 'exp'>,
+                             regressor=LinearRegression(copy_X=True,
+                                                        fit_intercept=True,
+                                                        n_jobs=1,
+                                                        normalize=False),
+                             transformer=None)
+    >>> tt.score(X, y)
+    1.0
+    >>> tt.regressor_.coef_
+    array([ 2.])
+
+    Notes
+    -----
+    Internally, the target ``y`` is always converted into a 2-dimensional array
+    to be used by scikit-learn transformers. At the time of prediction, the
+    output will be reshape to a have the same number of dimension than ``y``.
+
+    See :ref:`examples/preprocessing/plot_transform_target.py
+    <sphx_glr_auto_examples_preprocessing_plot_transform_target.py> `.
+
+    """
+    def __init__(self, regressor=None, transformer=None,
+                 func=None, inverse_func=None, check_inverse=True):
+        self.regressor = regressor
+        self.transformer = transformer
+        self.func = func
+        self.inverse_func = inverse_func
+        self.check_inverse = check_inverse
+
+    def _fit_transformer(self, y):
+        if (self.transformer is not None and
+                (self.func is not None or self.inverse_func is not None)):
+            raise ValueError("'transformer' and functions 'func'/"
+                             "'inverse_func' cannot both be set.")
+        elif self.transformer is not None:
+            self.transformer_ = clone(self.transformer)
+        else:
+            self.transformer_ = FunctionTransformer(
+                func=self.func, inverse_func=self.inverse_func, validate=True)
```

Comment:
Are we sure we want to validate? This is all that is excluding complex, for instance.

Reply:
For the complex, we will be block at fit time since we also have a check anyway.

```python
X, y = check_X_y(X, y, multi_output=True, y_numeric=True)
```

Basically this is why I turn validate to True. If we want to be permissive we need the transformer to be permissive all the way and avoid the check in all methods.

Reply:
Well, I think this is okay for most use-cases, so let's leave it as is. If someone wants it changed, they'll work out how to fix it...